test/functional/test_cluster_basic.py

```diff
@@ -104,6 +104,23 @@ def test_slave_reconnection_does_not_take_down_master(self):
         self._assert_build_completed_as_expected(build_1['build_id'], test_config, project_dir)
         self._assert_build_completed_as_expected(build_2['build_id'], test_config, project_dir)
 
+    def test_failed_setup_does_not_kill_slave(self):
+        master = self.cluster.start_master()
+        slave = self.cluster.start_slave()
+
+        project_dir = tempfile.TemporaryDirectory()
+        build_resp = master.post_new_build({
+            'type': 'directory',
+            'config': yaml.safe_load(FAILING_SETUP_JOB.config[os.name])['FailingSetupJob'],
+            'project_directory': project_dir.name,
+        })
+
+        build_id = build_resp['build_id']
+        self.assertTrue(master.block_until_build_finished(build_id, timeout=30),
+                        'The build should finish building within the timeout.')
+        self.assertTrue(slave.get_slave_status()['slave']['is_alive'],
+                        'Slave should not die even though setup failed')
```

Comment:
Do you want to assert that the build has an error status?

Testing the API response for the slave is not bad, but the real assertion would be that the slave is correctly marked idle on the master and is able to be assigned to other builds. We could add that to this test by asserting a subsequent build runs successfully.